File: third-party/pvm/pvm-3.4.6/pvm3/examples/spmd.c

static char rcsid[] =
	"$Id: spmd.c,v 1.5 2009/01/23 01:12:52 pvmsrc Exp $";

/*
 *         PVM version 3.4:  Parallel Virtual Machine System
 *               University of Tennessee, Knoxville TN.
 *           Oak Ridge National Laboratory, Oak Ridge TN.
 *                   Emory University, Atlanta GA.
 *      Authors:  J. J. Dongarra, G. E. Fagg, M. Fischer
 *          G. A. Geist, J. A. Kohl, R. J. Manchek, P. Mucci,
 *         P. M. Papadopoulos, S. L. Scott, and V. S. Sunderam
 *                   (C) 1997 All Rights Reserved
 *
 *                              NOTICE
 *
 * Permission to use, copy, modify, and distribute this software and
 * its documentation for any purpose and without fee is hereby granted
 * provided that the above copyright notice appear in all copies and
 * that both the copyright notice and this permission notice appear in
 * supporting documentation.
 *
 * Neither the Institutions (Emory University, Oak Ridge National
 * Laboratory, and University of Tennessee) nor the Authors make any
 * representations about the suitability of this software for any
 * purpose.  This software is provided ``as is'' without express or
 * implied warranty.
 *
 * PVM version 3 was funded in part by the U.S. Department of Energy,
 * the National Science Foundation and the State of Tennessee.
 */

/*
*    SPMD example using PVM 3
*    Uses pvm_siblings() to determine number of tasks that were spawned
*    together
*/

#include <stdio.h>
#ifdef HASSTDLIB
#include <stdlib.h>
#endif
#include <sys/types.h>
#include "pvm3.h"
#define MAXNPROC 32

void dowork();

main()
{
    int mytid;                  /* my task id */
    int *tids;                  /* array of task ids */
    int me;                     /* my process number */
    int i;
	int ntids;

    /* enroll in pvm */
    mytid = pvm_mytid();

	/* determine the size of my sibling list */

	ntids = pvm_siblings(&tids);

	for (i = 0; i < ntids; i ++)
		if ( tids[i] == mytid)
		{
			me = i;
			break;
		}
		
	if (me == 0)
	{
		printf("Pass a token through the %3d tid ring:\n", ntids);
		for (i = 0; i < ntids; i ++)
		{
			printf( "%6d -> ", tids[i]);
			if (i % 6 == 0 && i > 0)
				printf("\n");	
		}
		printf("%6d \n", tids[0]);
	}
/*--------------------------------------------------------------------------*/
     
     dowork( me, ntids, tids );

     /* program finished exit pvm */
     pvm_exit();
     exit(1);
}

/* Simple example passes a token around a ring */

void
dowork( me, nproc, tids )
     int me;
     int nproc;
	 int tids[];
{
     int token;
     int src, dest;
     int count  = 1;
     int stride = 1;
     int msgtag = 4;

     /* Determine neighbors in the ring */
     if ( me == 0 )
		src = tids[nproc -1];
	 else
 		src = tids[me -1];

     if (me == nproc - 1)
        dest = tids[0];
     else
        dest = tids[me + 1];

     if( me == 0 )
     { 
        token = dest;
        pvm_initsend( PvmDataDefault );
        pvm_pkint( &token, count, stride );
        pvm_send( dest, msgtag );
        pvm_recv( src, msgtag );
        printf("token ring done\n");
     }
     else
     {
        pvm_recv( src, msgtag );
        pvm_upkint( &token, count, stride );
        pvm_initsend( PvmDataDefault );
        pvm_pkint( &token, count, stride );
        pvm_send( dest, msgtag );
     }
}

File: third-party/pvm/pvm-3.4.6/pvm3/examples/spmd.f
c
c $Id: spmd.f,v 1.4 1997/08/28 20:11:44 pvmsrc Exp $
c
c---------------------------------------- 
c   SPMD Fortran example using PVM 3
c   Illustrates use of new pvm3.4 call pvmfsiblings
c---------------------------------------- 
       program spmd
       include '../include/fpvm3.h'
       PARAMETER( MAXNPROC=32 )

       integer mytid, me, info
       integer tids(0:MAXNPROC -1)
       integer ntids 

c      -------------
c      Enroll in pvm
c      -------------
       call pvmfmytid( mytid )

c      --------------------------------------------
c      Call pvmfsiblings to determine how many tasks were 
c      spawned with me. 
c      --------------------------------------------
       me = -1 
       call pvmfsiblings(ntids, 0, tids(0))
       if (ntids .gt. MAXNPROC) ntids = MAXNPROC
	 
       do i = 0, ntids - 1
           call pvmfsiblings(ntids, i, tids(i))
           if (tids(i) .eq. mytid) me = i
       end do
	

       if (me .eq. -1) then
           call pvmfexit(info)
           stop
       endif
      
       if (me .eq. 0) then
          write (6,*) 'Pass a token through the', ntids, ' tid ring:'
		  write (6,6000) (tids(i), i=0, ntids-1), tids(0)
 6000     format( 6(I7:, ' -> '))
       end if
		
       call dowork( me, ntids )

c      -------------------------
c      program finished exit pvm
c      -------------------------
       call pvmfexit(info)
       stop
       end


       subroutine dowork( me, nproc )
       include '../include/fpvm3.h'
c-------------------------------------------------
c Simple subroutine to pass a token around a ring
c-------------------------------------------------
       integer me, nproc
 
       integer token, src, dest, count, stride, msgtag 
       integer ndum
c      -------------------------------
c      Determine neighbors in the ring
c      -------------------------------
       call pvmfsiblings(ndum, me-1, src )
       call pvmfsiblings(ndum, me+1, dest )
       if( me .eq. 0 ) call pvmfsiblings( ndum, nproc-1, src )
       if( me .eq. nproc - 1 ) call pvmfsiblings( ndum, 0, dest)

       count  = 1
       stride = 1
       msgtag = 4


       if( me .eq. 0 ) then
          token = dest
          call pvmfinitsend( PVMDEFAULT, info )
          call pvmfpack( INTEGER4, token, count, stride, info )
          call pvmfsend( dest, msgtag, info )
          call pvmfrecv( src, msgtag, info )
          print*, 'token ring done'
       else
          call pvmfrecv( src, msgtag, info )
          call pvmfunpack( INTEGER4, token, count, stride, info )
          call pvmfinitsend( PVMDEFAULT, info )
          call pvmfpack( INTEGER4, token, count, stride, info )
          call pvmfsend( dest, msgtag, info )
       endif
      
       return
       end
